### PR TITLE
P4-OVS: Introduce support for tap port creation

### PR DESCRIPTION
**Pull request by nupuruttarwar**
This patch introduces support required for tap port creation and
GNMI cli enhancements for pipeline name, mtu value and mempool
name DPDK port configuration.

Example commands:
With all parameters: sudo gnmi-cli set "device:virtual-device,\
name:TAP0,pipeline-name:pipe,mempool-name:MEMPOOL0,mtu:1500,port-type:TAP"
Default parameters skipped: sudo gnmi-cli set "device:virtual-device,\
name:TAP1,mtu:1500,port-type:TAP"

Title: Introduce support for tap port creation
Change-type: FeatureImplementation
Signed-off-by: nupuruttarwar <nupur.uttarwar@intel.com>

### CODE REVIEW

**Comment by stolsma** [comment]
Nice! Will have a look after the weekend but don't wait for me if you want to merge!👍

File: external/PATCH-01-STRATUM
Comment (stolsma) [inline external/PATCH-01-STRATUM:60]: Could all the formatting changes not related to implementing TAP support be included in another PR? It is very difficult to see which code changed and to get to understand relations between code when there is so much clutter....

File: external/PATCH-01-STRATUM
Comment (stolsma) [inline external/PATCH-01-STRATUM:722]: Does the change from LINK to VHOST naming also means the code and documentation in IPDK need to be changed?

File: p4proto/bfIntf/bf_chassis_manager.cc
Comment (stolsma) [inline p4proto/bfIntf/bf_chassis_manager.cc:203]: Also LINK to VHOST question related to operational use...

File: p4proto/bfIntf/bf_chassis_manager.h
Comment (stolsma) [inline p4proto/bfIntf/bf_chassis_manager.h:42]: Why use a negative mask instead of defining a seperate GNMI_CONFIG_TAP mask defining the TAP parameters? That will make it easier to add other link types later on with other specific parameters if needed ?

File: utilities/gnmi_cli.cc
Comment (stolsma) [inline utilities/gnmi_cli.cc:176]: Will this `virtual-device` and  `physical-device` option be documented somewhere? Or is it already documented and I don't know it exists?

**Review by stolsma — CHANGES_REQUESTED** [review]
Looks good in general, I have mostly some questions about changing cli use and documenting it. Can't wait until it is merged! 😃

File: external/PATCH-01-STRATUM
Comment (nupuruttarwar) [reply external/PATCH-01-STRATUM:60]: @stolsma thanks for the review!! Since it was a minor change, I added it here. But you are right, I can create a separate PR for such changes from next time

File: external/PATCH-01-STRATUM
Comment (nupuruttarwar) [reply external/PATCH-01-STRATUM:722]: That's not required. Internal logic introduced on line 284 in gnmi_cli.cc file will change the port type to vhost if device-type is virtual device and port type is link. This is to avoid confusion for existing users. Also, both vhost user ports and pci link ports are treated as link ports by DPDK backend. I am working on documentation for tap ports, will update the PR soon

File: utilities/gnmi_cli.cc
Comment (nupuruttarwar) [reply utilities/gnmi_cli.cc:176]: Will remove physical device since this is for pci based ports, support for which is being still worked upon. Once this support is available, will surely add the documentation :)

File: p4proto/bfIntf/bf_chassis_manager.cc
Comment (nupuruttarwar) [reply p4proto/bfIntf/bf_chassis_manager.cc:203]: Sorry, I didn't understand the question here?

**Review by Namrata-intel — APPROVED** [review]

**Pull request merged**
